### PR TITLE
Capture unexpected caching disabled reasons

### PR DESCRIPTION
**Pull request by blindpirate**
For example, "OVERLAPPING_OUTPUTS".

In this way, we can set up a monitor for such events.

### CODE REVIEW

File: ge-export/src/test/kotlin/org/gradle/devprod/collector/enterprise/export/extractor/UnexpectedCachingDisableReasonsExtractorTest.kt [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package org.gradle.devprod.collector.enterprise.export.extractor
+
+import org.junit.jupiter.api.Assertions.assertEquals
+import org.junit.jupiter.api.Test
+
+class UnexpectedCachingDisableReasonsExtractorTest {
+    @Test
+    fun `can extract unexpected caching disabled reasons`() {
+        val events = parse("/ngapnu6jwslkk.txt")
```
Comment (hegyibalint) [inline ge-export/src/test/kotlin/org/gradle/devprod/collector/enterprise/export/extractor/UnexpectedCachingDisableReasonsExtractorTest.kt:9]: Such a weird filename, was this a real output?

**Review by hegyibalint — APPROVED** [review]

File: ge-export/src/test/kotlin/org/gradle/devprod/collector/enterprise/export/extractor/UnexpectedCachingDisableReasonsExtractorTest.kt (hunk 1 above)
Comment (blindpirate) [reply ge-export/src/test/kotlin/org/gradle/devprod/collector/enterprise/export/extractor/UnexpectedCachingDisableReasonsExtractorTest.kt:9]: Yes, this is build scan id.

File: ge-export/src/main/kotlin/org/gradle/devprod/collector/enterprise/export/extractor/extractors.kt [hunk 2]
```diff
@@ -256,6 +256,16 @@ object DaemonUnhealthy : SingleEventExtractor<String?>("DaemonUnhealthy") {
     }
 }
 
+// https://docs.gradle.com/enterprise/event-model-javadoc/com/gradle/scan/eventmodel/gradle/TaskFinished_1_6.html#cachingDisabledReason
+object UnexpectedCachingDisableReasonsExtractor : SingleEventExtractor<List<String>>("TaskFinished") {
+    override fun extract(events: Iterable<BuildEvent>): List<String> {
+        return events
+            .mapNotNull { it.data?.stringProperty("cachingDisabledReasonCategory") }
+            .filter { it == "OVERLAPPING_OUTPUTS" }
```
Comment (wolfs) [inline ge-export/src/main/kotlin/org/gradle/devprod/collector/enterprise/export/extractor/extractors.kt:264]: 🤔 How about we capture all caching disabled reasons? Do you think that would blow up the table?

Comment (blindpirate) [reply ge-export/src/main/kotlin/org/gradle/devprod/collector/enterprise/export/extractor/extractors.kt:264]: That's probably not a big issue... we run ~10K builds per day, and all possible value of `cachingDisabledReasonCategory` are:
 
- [BUILD_CACHE_NOT_ENABLED](https://docs.gradle.com/enterprise/event-model-javadoc/com/gradle/scan/eventmodel/gradle/task/TaskOutputCachingDisabledReason_1.html#BUILD_CACHE_NOT_ENABLED)	 
- [MULTIPLE_OUTPUTS_DECLARED](https://docs.gradle.com/enterprise/event-model-javadoc/com/gradle/scan/eventmodel/gradle/task/TaskOutputCachingDisabledReason_1.html#MULTIPLE_OUTPUTS_DECLARED)	 
- [NO_OUTPUTS_DECLARED](https://docs.gradle.com/enterprise/event-model-javadoc/com/gradle/scan/eventmodel/gradle/task/TaskOutputCachingDisabledReason_1.html#NO_OUTPUTS_DECLARED)	 
- [OUTPUT_CACHING_NOT_ENABLED](https://docs.gradle.com/enterprise/event-model-javadoc/com/gradle/scan/eventmodel/gradle/task/TaskOutputCachingDisabledReason_1.html#OUTPUT_CACHING_NOT_ENABLED)	 
- [OVERLAPPING_OUTPUTS](https://docs.gradle.com/enterprise/event-model-javadoc/com/gradle/scan/eventmodel/gradle/task/TaskOutputCachingDisabledReason_1.html#OVERLAPPING_OUTPUTS)	 
- [SATISFIED_PREDICATE_DO_NOT_CACHE_IF](https://docs.gradle.com/enterprise/event-model-javadoc/com/gradle/scan/eventmodel/gradle/task/TaskOutputCachingDisabledReason_1.html#SATISFIED_PREDICATE_DO_NOT_CACHE_IF)	 
- [UNKNOWN](https://docs.gradle.com/enterprise/event-model-javadoc/com/gradle/scan/eventmodel/gradle/task/TaskOutputCachingDisabledReason_1.html#UNKNOWN)	 
- [UNSATISFIED_PREDICATE_CACHE_IF](https://docs.gradle.com/enterprise/event-model-javadoc/com/gradle/scan/eventmodel/gradle/task/TaskOutputCachingDisabledReason_1.html#UNSATISFIED_PREDICATE_CACHE_IF)

Suppose each build has 3~4 reasons, that would be 10K * ~100 B = ~1MB per day. But I think capturing other reasons doesn't make much sense - if we like, we can capture more on demand in the future.

**Comment by blindpirate** [comment]
I deployed this branch and it seems to be working:

image

File: ge-export/src/main/kotlin/org/gradle/devprod/collector/enterprise/export/extractor/extractors.kt (hunk 2 above)
Comment (blindpirate) [reply ge-export/src/main/kotlin/org/gradle/devprod/collector/enterprise/export/extractor/extractors.kt:264]: Another reason for only capturing "unexpected" disabling reason is that we can avoid hardcoding the alert rule:

```
select * from build where unexpected_caching_disabled_reason.isNotEmpty()
```

In the future if we want to monitor other reasons, the alert rule doesn't need to be changed.

**Pull request merged**